operator/cilium_node.go

```diff
@@ -1,112 +1,428 @@
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err != nil {
+					log.WithError(err).Warning("Unable to process CiliumNode Add event")
+					return
 				}
+				queue.Add(key)
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				if oldNode := k8s.ObjToCiliumNode(oldObj); oldNode != nil {
-					if node := k8s.ObjToCiliumNode(newObj); node != nil {
-						if oldNode.DeepEqual(node) {
+					if newNode := k8s.ObjToCiliumNode(newObj); newNode != nil {
+						if oldNode.DeepEqual(newNode) {
 							return
 						}
-						// node is deep copied before it is stored in pkg/aws/eni
-						nodeManager.Update(node)
+						key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(newObj)
+						if err != nil {
+							log.WithError(err).Warning("Unable to process CiliumNode Update event")
+							return
+						}
+						queue.Add(key)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", newNode, newNode)
 					}
+				} else {
+					log.Warningf("Unknown CiliumNode object type %T received: %+v", oldNode, oldNode)
 				}
 			},
 			DeleteFunc: func(obj interface{}) {
-				if node := k8s.ObjToCiliumNode(obj); node != nil {
-					nodeManager.Delete(node.Name)
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err != nil {
+					log.WithError(err).Warning("Unable to process CiliumNode Delete event")
+					return
 				}
+				queue.Add(key)
 			},
-		},
-		k8s.ConvertToCiliumNode,
+		}
+	} else {
+		// Since we won't be handling any events we don't need to convert
+		// objects.
+		ciliumNodeConvertFunc = nil
+	}
+
+	// TODO: The operator is currently storing a full copy of the
+	// CiliumNode resource, as the resource grows, we may want to consider
+	// introducing a slim version of it.
+	var ciliumNodeInformer cache.Controller
+	ciliumNodeStore, ciliumNodeInformer = informer.NewInformer(
+		cache.NewListWatchFromClient(ciliumK8sClient.CiliumV2().RESTClient(),
+			cilium_v2.CNPluralName, v1.NamespaceAll, fields.Everything()),
+		&cilium_v2.CiliumNode{},
+		0,
+		resourceEventHandler,
+		ciliumNodeConvertFunc,
 	)
 
 	go func() {
 		cache.WaitForCacheSync(wait.NeverStop, ciliumNodeInformer.HasSynced)
 		close(k8sCiliumNodesCacheSynced)
+		// Only handle events if syncHandler is not nil. If it is nil then
+		// there isn't any event handler set for CiliumNodes events.
+		if syncHandler != nil {
+			for processNextWorkItem(queue, syncHandler) {
+			}
+		}
 	}()
 
 	go ciliumNodeInformer.Run(wait.NeverStop)
+
+	return nil
 }
 
-func deleteCiliumNode(nodeManager *allocator.NodeEventHandler, name string) {
-	if nodeManager != nil {
-		(*nodeManager).Delete(name)
+// processNextWorkItem process all events from the workqueue.
+func processNextWorkItem(queue workqueue.RateLimitingInterface, syncHandler func(key string) error) bool {
+	// Get next event from the queue
+	key, quit := queue.Get()
+	if quit {
+		return false
+	}
+	// Mark it as Done
```

Comment:
nit: I suspect that these were added in response to my earlier feedback, but a couple of these comments are superfluous, the code `queue.Get()` clearly gets the next event from the queue. Similarly the `queue.Done()` call marks it as done, that's clear from the code. If that's how the workqueue interface works, we don't have a choice and the comments don't help explain why, they just duplicate the explanation from the code in different words. I think it's fine to drop those comments.

```suggestion
	key, quit := queue.Get()
	if quit {
		return false
	}
```